sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java

```diff
@@ -468,66 +440,174 @@ private static Expression value(
+          }
+          break;
+        case MAP:
+          value = Expressions.call(expression, "getMap", Expressions.constant(index));
+          break;
+        case ROW:
+          value = Expressions.call(expression, "getRow", Expressions.constant(index));
+          break;
+        case LOGICAL_TYPE:
+          String identifier = fieldType.getLogicalType().getIdentifier();
+          if (CharType.IDENTIFIER.equals(identifier)) {
+            value = Expressions.call(expression, "getString", Expressions.constant(index));
+          } else if (TimeWithLocalTzType.IDENTIFIER.equals(identifier)) {
+            value = Expressions.call(expression, "getDateTime", Expressions.constant(index));
+          } else if (SqlTypes.DATE.getIdentifier().equals(identifier)) {
+            value =
+                Expressions.convert_(
+                    Expressions.call(
+                        expression,
+                        "getLogicalTypeValue",
+                        Expressions.constant(index),
+                        Expressions.constant(LocalDate.class)),
+                    LocalDate.class);
+          } else if (SqlTypes.TIME.getIdentifier().equals(identifier)) {
+            value =
+                Expressions.convert_(
+                    Expressions.call(
+                        expression,
+                        "getLogicalTypeValue",
+                        Expressions.constant(index),
+                        Expressions.constant(LocalTime.class)),
+                    LocalTime.class);
+          } else if (SqlTypes.DATETIME.getIdentifier().equals(identifier)) {
+            value =
+                Expressions.convert_(
+                    Expressions.call(
+                        expression,
+                        "getLogicalTypeValue",
+                        Expressions.constant(index),
+                        Expressions.constant(LocalDateTime.class)),
+                    LocalDateTime.class);
+          } else {
+            throw new UnsupportedOperationException("Unable to get logical type " + identifier);
+          }
+          break;
+        default:
+          throw new UnsupportedOperationException("Unable to get " + fieldType.getTypeName());
       }
 
-      Expression value =
-          Expressions.convert_(
-              Expressions.call(
-                  expression,
-                  "getBaseValue",
-                  Expressions.constant(index),
-                  Expressions.constant(convertTo)),
-              convertTo);
-      return (storageType != Object.class) ? value(value, fromType) : value;
+      return value(value, fieldType);
     }
 
-    private static Expression value(Expression value, Schema.FieldType type) {
-      if (type.getTypeName().isLogicalType()) {
-        String logicalId = type.getLogicalType().getIdentifier();
-        if (SqlTypes.TIME.getIdentifier().equals(logicalId)) {
+    private static Expression value(Expression value, FieldType fieldType) {
```

Comment:
Can we rename this method to differentiate it from `value(BlockBuilder list, int index, Type storageType, Expression input, Schema schema)` and better describe what it's actually doing?

Reply:
Yep, renamed to getBeamField